xarray/coding/times.py

```diff
@@ -1123,14 +1122,39 @@ def _eagerly_encode_cf_datetime(
         num = reshape(num.values, dates.shape)
 
     except (OutOfBoundsDatetime, OverflowError, ValueError):
+        time_units, ref_date = _unpack_time_units_and_ref_date_cftime(units, calendar)
+        time_delta_cftime = _time_units_to_timedelta(time_units)
+        needed_units = _infer_needed_units_cftime(ref_date, data_units, calendar)
+        needed_time_delta_cftime = _time_units_to_timedelta(needed_units)
+
+        if (
+            np.issubdtype(dtype, np.integer)
+            and time_delta_cftime > needed_time_delta_cftime
+        ):
+            new_units = f"{needed_units} since {format_cftime_datetime(ref_date)}"
+            if allow_units_modification:
+                units = new_units
+                emit_user_level_warning(
+                    f"Times can't be serialized faithfully to int64 with requested units {units!r}. "
+                    f"Serializing with units {new_units!r} instead. "
+                    f"Set encoding['dtype'] to floating point dtype to serialize with units {units!r}. "
+                    f"Set encoding['units'] to {new_units!r} to silence this warning ."
+                )
```

Comment:
The assignment needs to go after the warning, otherwise the warning would return same for "units" and "new_units".

```suggestion
                emit_user_level_warning(
                    f"Times can't be serialized faithfully to int64 with requested units {units!r}. "
                    f"Serializing with units {new_units!r} instead. "
                    f"Set encoding['dtype'] to floating point dtype to serialize with units {units!r}. "
                    f"Set encoding['units'] to {new_units!r} to silence this warning ."
                )
                units = new_units
```

Reply:
Important catch—thanks!